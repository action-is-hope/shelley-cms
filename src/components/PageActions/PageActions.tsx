import React, { forwardRef, Ref, ReactElement } from "react";
import type { CollectionChildren } from "@react-types/shared/src/collections";
import { st, classes } from "./pageActions.st.css";
import {
  Text,
  Button,
  Icon,
  MenuTrigger,
  Menu,
  ButtonGroup,
  MenuTriggerProps,
  ComponentBase,
} from "@actionishope/shelley";

import { classes as spacing } from "@actionishope/shelley/styles/default/spacing.st.css";

export type statusOptions =
  | "published"
  | "draft"
  | "updated"
  | "archived"
  | "unpublished";

export interface PageActionsProps<T>
  extends Omit<React.HTMLAttributes<HTMLDivElement>, "children">,
    ComponentBase {
  /** Provide a status */
  status: statusOptions;
  /** The last saved value, e.g "a week ago", "a few minutes" etc. */
  lastSaved?: string;
  /**
   * Changes the button into a "Review required" state with
   * onAction responding with "review"
   */
  reviewRequired?: boolean;
  /** An array of disable keys to disable menu items. */
  disabledKeys?: string[];
  /**
   * The `onAction` event is fired when either the main
   * button or one of the menu options is selected.
   * See @docs for more info.
   */
  onAction: (key: string | number) => void;
  /** Items for the Menu drop down. */
  children: CollectionChildren<T>;
  /** Position props for the internal Menu. */
  position?: Pick<
    MenuTriggerProps,
    "placement" | "shouldFlip" | "offset" | "crossOffset" | "portalSelector"
  >;
}

function PageActions<T extends { key: string }>(
  props: PageActionsProps<T>,
  ref?: React.Ref<HTMLDivElement>
) {
  const {
    className: classNameProp,
    children,
    status = "draft",
    lastSaved,
    reviewRequired,
    disabledKeys: disabledKeysProp,
    position,
    onAction: onActionProp,
    "data-id": dataId,
    ...rest
  } = props;

  const strings = {
    status: "Status:",
    publish: "Publish",
    unpublish: "Unpublish",
    draft: "Draft",
    archive: "Archive",
    unarchive: "Unarchive",
    publishChanges: "Publish changes",
    publishMenu: "Publish menu",
    published: "Published",
    updated: "Updated",
    archived: "Archived",
    unpublished: "Unpublished",
    reviewRequired: "Review required",
    delete: "Delete",
    changeStatus: "Change status",
    changesSaved: "Changes saved:",
  };

  const lookUp = {
    updated: {
      text: strings.publishChanges,
      action: "publish",
      disabled: false,
    },
    published: {
      text: strings.publish,
      action: "none",
      disabled: true,
    },
    unpublished: {
      text: strings.publish,
      action: "publish",
      disabled: false,
    },
    draft: {
      text: strings.publish,
      action: "publish",
      disabled: false,
    },
    archived: {
      text: strings.unarchive,
      action: "unarchive",
      disabled: false,
    },
  };

  // const disabledKeys = Children.map(children, (child) => {
  //   if (child?.key === "archive" && status === "archived") {
  //     return child?.key;
  //   }
  //   if (child?.key === "unpublish" && status === "unpublished") {
  //     return child?.key;
  //   }
  //   if (child?.key === "make-draft" && status === "draft") {
  //     return child?.key;
  //   }
  // });

  return (
    <div
      className={st(classes.root, classNameProp)}
      {...rest}
      ref={ref}
      data-id={dataId}
    >
      <div className={classes.info}>
        <Text
          as="div"
          vol={1}
          className={st(spacing.mb1, classes.lastSaved)}
          data-id={dataId ? `${dataId}--last-saved` : undefined}
        >
          {lastSaved && `${strings.changesSaved} ${lastSaved}`}
        </Text>
        {/* extra children? */}
      </div>

      <div className={classes.actions}>
        <Text
          as="div"
          vol={1}
          className={st(spacing.mb1, classes.statusText)}
          data-id={dataId ? `${dataId}--status` : undefined}
        >
          {/* @todo: Convert to StatusIndicator component */}
          <span className={st(classes.led, classes[status])}></span>
          <strong>{strings.status}</strong> {strings[status]}
        </Text>
        {/* @todo: shelley -> Button group us adding extra props to the MenuTrigger in this scenario */}
        <ButtonGroup className={classes.buttonGroup} splitButton>
          <Button
            vol={4}
            variant="primary"
            tone={reviewRequired ? 2 : 1}
            className={classes.reviewButton}
            data-id={dataId ? `${dataId}--actionButton` : undefined}
            fullWidth
            onPress={() =>
              onActionProp(reviewRequired ? "review" : lookUp[status].action)
            }
            isDisabled={reviewRequired ? false : lookUp[status].disabled}
          >
            {reviewRequired ? strings.reviewRequired : lookUp[status].text}
          </Button>
          <MenuTrigger offset={8} crossOffset={-80} {...position}>
            <Button
              icon={
                <Icon alt={strings.publishMenu}>
                  <path d="M13 4v2l-5 5-5-5v-2l5 5z"></path>
                </Icon>
              }
              className={classes.menuTrigger}
              data-id={dataId ? `${dataId}--menuTrigger` : undefined}
              vol={4}
              tone={1}
            />
            <Menu
              onAction={(key) => onActionProp(key)}
              // disabledKeys={disabledKeys}
              disabledKeys={disabledKeysProp}
              className={classes.menu}
              data-id={dataId ? `${dataId}--menu` : undefined}
            >
              {children}
            </Menu>
          </MenuTrigger>
        </ButtonGroup>
      </div>
    </div>
  );
}

// forwardRef doesn't support generic parameters -> cast to the correct type.
// https://stackoverflow.com/questions/58469229/react-with-typescript-generics-while-using-react-forwardref
const _PageActions = forwardRef(PageActions) as <T>(
  props: PageActionsProps<T> & { ref?: Ref<HTMLDivElement> }
) => ReactElement;
export { _PageActions as PageActions };
