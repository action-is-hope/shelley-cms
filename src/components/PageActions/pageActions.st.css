/* pageActions.st.css */
@namespace "PageActions";
@st-import Button from "@actionishope/shelley/components/Button/button.st.css";
@st-import ButtonGroup from "@actionishope/shelley/components/ButtonGroup/buttonGroup.st.css";
@st-import Menu from "@actionishope/shelley/components/Menu/menu.st.css";

.root {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-column-gap: 24px;
  grid-row-gap: 4px;
  padding: 14px 24px 24px 8px;
}

.info {}

.lastSaved {}

.actions {}

.statusText {}

.reviewButton {
  -st-extends: Button;
}

.menuTrigger {
  -st-extends: Button;
}

.menu {
  -st-extends: Menu;
}

.buttonGroup {
  -st-extends: ButtonGroup;
  width: 100%;
}

/* @todo: Move to StatusIndicator comp */
.led {
  margin: -2px 8px 0 0;
  width: 12px;
  height: 12px;
  display: inline-block;
  border-radius: 50%;
  vertical-align: middle;
}

.archived {
  background-color: #f05751;
  border: 1px solid #d9453f;
}

.draft {
  background-color: #fba012;
  border: 1px solid #db8500;
}

.unpublished {
  background-color: #fba012;
  border: 1px solid #db8500;
}

.published {
  background-color: #14d997;
  border: 1px solid #0eb87f;
}

.updated {
  background-color: #15c8e8;
  border: 1px solid #75eaff;
}

/* 
@todo: Potentially add animation to the publlish button.
.reviewButton {
  -st-extends: Button;
  background-color: #222;
  animation-name: color;
  animation-duration: 3s;
  animation-iteration-count: infinite;
}

@keyframes color {
  0% {
    background-color: #16a78e;
  }

  50% {
    background-color: #448688;
  }

  100% {
    background-color: #16a78e;
  }
} */
