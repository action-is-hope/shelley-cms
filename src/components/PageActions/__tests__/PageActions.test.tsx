import { PageActions } from "../PageActions";
import renderer from "react-test-renderer";
import { Item } from "@actionishope/shelley";
import { SSRProvider } from "react-aria";

describe("PageActions", () => {
  it("renders page actions block with custom className", () => {
    const tree = renderer
      .create(
        // Internal use of useID breaks snapshots so using the SSRProvider.
        <SSRProvider>
          <PageActions
            className={"custom-classname"}
            lastSaved="5 mins ago"
            // reviewRequired={reviewRequired}
            status={"draft"}
            onAction={(action) => console.log(action)}
          >
            <Item key="archive">Archive</Item>
            <Item key="unpublish">Unpublish</Item>
          </PageActions>
        </SSRProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
