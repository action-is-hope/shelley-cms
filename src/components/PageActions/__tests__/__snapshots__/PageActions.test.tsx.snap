// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageActions renders page actions block with custom className 1`] = `
<div
  className="PageActions__Test__root custom-classname"
>
  <div
    className="PageActions__Test__info"
  >
    <div
      className="Text__Test__root Text__Test---vol-1-1 DefaultSpacing__Test__mb1 PageActions__Test__lastSaved"
    >
      Changes saved: 5 mins ago
    </div>
  </div>
  <div
    className="PageActions__Test__actions"
  >
    <div
      className="Text__Test__root Text__Test---vol-1-1 DefaultSpacing__Test__mb1 PageActions__Test__statusText"
    >
      <span
        className="PageActions__Test__led PageActions__Test__draft"
      />
      <strong>
        Status:
      </strong>
       
      Draft
    </div>
    <div
      className="ButtonGroup__Test__root ButtonGroup__Test---orientation-10-horizontal ButtonGroup__Test--splitButton PageActions__Test__buttonGroup"
    >
      <button
        className="Button__Test__root Button__Test--fullWidth Button__Test---tone-1-1 Button__Test---variant-7-primary Button__Test---vol-1-4 PageActions__Test__reviewButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        type="button"
      >
        <span
          className="Button__Test__inner"
        >
          Publish
        </span>
      </button>
      <button
        aria-controls={null}
        aria-expanded={false}
        aria-haspopup={true}
        className="Button__Test__root Button__Test---tone-1-1 Button__Test---variant-7-primary Button__Test---vol-1-4 PageActions__Test__menuTrigger"
        id="react-aria-1"
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        type="button"
      >
         
        <span
          className="VisuallyHidden__Test__root"
        >
          Publish menu
        </span>
        <svg
          aria-hidden={true}
          className="Icon__Test__root"
          viewBox="0 0 16 16"
        >
          <path
            d="M13 4v2l-5 5-5-5v-2l5 5z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
